mito-ai/src/Extensions/AiChat/ChatTaskpane.tsx

```diff
@@ -94,68 +115,143 @@ const ChatTaskpane: React.FC<IChatTaskpaneProps> = ({
     // unlike state variables, which are captured at the beginning of a function and may not reflect updates made during execution.
     const shouldContinueAgentExecution = useRef<boolean>(true);
 
-    const fetchInitialChatHistory = async (): Promise<OpenAI.Chat.ChatCompletionMessageParam[]> => {
-        await websocketClient.ready;
+    const fetchChatThreads = async () => {
+      await websocketClient.ready;
 
-        const fetchHistoryCompletionRequest: IFetchHistoryCompletionRequest = {
-            type: 'fetch_history',
-            message_id: UUID.uuid4(),
-            metadata: {
-                promptType: 'fetch_history'
-            },
-            stream: false
+      const metadata: IGetThreadsMetadata = {
+        promptType: "get_threads"
+      };
+
+      const chatThreadsResponse = await websocketClient.sendMessage<
+        ICompletionRequest, 
+        IFetchThreadsReply
+      >({
+         type: "get_threads",
+         message_id: UUID.uuid4(),
+         metadata: metadata,
+         stream: false
+      });
+
+      setChatThreads(chatThreadsResponse.threads);
+    };
+
+    const fetchChatHistoryForThread = async (threadId: string) => {
+      await websocketClient.ready;
+
+      const metadata: IFetchHistoryMetadata = {
+        promptType: "fetch_history",
+        threadID: threadId
+      };
+
+      const fetchHistoryCompletionRequest: IFetchHistoryCompletionRequest = {
+        type: 'fetch_history',
+        message_id: UUID.uuid4(),
+        metadata: metadata,
+        stream: false
+    }
+
+      const chatHistoryResponse = await websocketClient.sendMessage<
+        ICompletionRequest, 
+        IFetchHistoryReply
+      >(fetchHistoryCompletionRequest);
+
+      // Create a fresh ChatHistoryManager and add the initial messages
+      const newChatHistoryManager = getDefaultChatHistoryManager(
+        notebookTracker,
+        contextManager
+    );
+
+      // Add messages to the ChatHistoryManager
+      chatHistoryResponse.items.forEach(item => {
+        try {
+            // If the user sent a message in agent mode, the ai response will be a JSON object
+            // which we need to parse. 
+            const agentResponse = JSON.parse(item.content as string);
+            handleAgentResponse(agentResponse, newChatHistoryManager);
+        } catch {
+            newChatHistoryManager.addChatMessageFromHistory(item);
         }
-        
-        const chatHistoryResponse = await websocketClient.sendMessage(fetchHistoryCompletionRequest);
+      });
 
-        return chatHistoryResponse.items.map((item: any) => ({
-            role: item.role,
-            content: item.content
-        }));
+      // Update the state with the new ChatHistoryManager
+      setChatHistoryManager(newChatHistoryManager);
+      setActiveThreadId(threadId);
+    };
+    
+    const deleteThread = async (threadId: string) => {
+      await websocketClient.ready;
```

Comment:
Do we need this at the start of each function? Instead can we move it into the sendMessage function so we don't have to remember to do it everywhere?